Source/bmalloc/libpas/src/libpas/pas_enumerator.c

```diff
@@ -95,39 +95,52 @@ pas_enumerator* pas_enumerator_create(pas_root* remote_root_address,
         result, sizeof(void*) * pas_heap_config_kind_num_kinds);
     pas_zero_memory(result->heap_config_datas, sizeof(void*) * pas_heap_config_kind_num_kinds);
 
-    result->root = reader(result, remote_root_address, sizeof(pas_root), reader_arg);
-    if (!result->root)
+    /*
+     * Some non-obvious behavior: even if we expect the remote address of a value (like the
+     * remote pas_root) to be persistent, our local mapping of the remote memory is allowed to
+     * change, and can in fact change any time we call the reader. So, it's unsafe to store
+     * a pointer to the remote memory across a call to reader, and we need to instead allocate
+     * local storage for these values and copy the remote values in.
+     */
+
+    pas_root* remote_root = reader(result, remote_root_address, sizeof(pas_root), reader_arg);
```

Comment:
Is this (declaring a local var in the middle of the function) allowed by C?

Reply:
Yes as of C99 or later, which I'm pretty sure we support...but _most_ of libPAS respects this, so I'll refactor it real quick.

Reply:
Resolved in latest push.